src/core/AutomationPattern.cpp

```diff
@@ -502,74 +497,84 @@ void AutomationPattern::flipY()
 
 
 
-void AutomationPattern::flipX( int length )
+void AutomationPattern::flipX(int length)
 {
+	timeMap::const_iterator it = m_timeMap.lowerBound(0);
+	if(it == m_timeMap.end())
+	{
+		return;
+	}
+
+	// Temporary map where we will store the flipped version
+	// of our pattern
 	timeMap tempMap;
 
-	timeMap::const_iterator iterate = m_timeMap.lowerBound(0);
 	float tempValue = 0;
 	float tempOutValue = 0;
-	int numPoints = 0;
-
-	// TODO: Same as the comment on AutomationPattern::flipY
-	for( int i = 0; ( iterate + i + 1 ) != m_timeMap.end() && ( iterate + i ) != m_timeMap.end() ; i++)
-	{
-		numPoints++;
-	}
 
-	float realLength = ( iterate + numPoints ).key();
+	// We know the QMap isn't empty, making this safe:
+	float realLength = m_timeMap.lastKey();
 
-	if ( length != -1 && length != realLength)
+	// If we have a positive length, we want to flip the area covered by that
+	// length, even if it goes beyond the pattern. A negative length means that
+	// we just want to flip the nodes we have
+	if (length >= 0 && length != realLength)
 	{
 		// If length to be flipped is bigger than the real length
-		if ( realLength < length )
+		if (realLength < length)
 		{
-			// Adds a node to the end of the region being flipped because the region flipped goes
-			// further than the last node
-			tempValue = valueAt( ( iterate + numPoints ).key() );
-			putValue( MidiTime( length ) , tempValue, false);
-			numPoints++;
+			// We are flipping an area that goes beyond the last node. So we add a node to the
+			// beginning of the flipped timeMap representing the value of the end of the area
+			tempValue = valueAt(realLength);
+			tempMap[0] = AutomationNode(this, tempValue, 0);
 
-			for( int i = 0; i <= numPoints; i++ )
+			// Now flip the nodes we have in relation to the length
+			while (it != m_timeMap.end())
```

Comment:
All these whiles could be changed to do while since the first if guards from this being the end. Optional of course, but just an observation.